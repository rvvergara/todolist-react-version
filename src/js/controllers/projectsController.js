import Project from '../models/project';
import * as localStorageData from './helpers/common/storage';

const projectsController = (
  () => {
    const projectsArray = localStorageData.getDataFromLocalStorage('projectsArray') ? localStorageData.getDataFromLocalStorage('projectsArray') : [];
    if (localStorageData.getDataFromLocalStorage('projectCount') === null) {
      localStorageData.setDataIntoLocalStorage('projectCount', 0);
    }
    return {
      create(name, id) {
        if (name === '') name = 'Project Name (Please customize name)';
        let projectCount = localStorageData.getDataFromLocalStorage('projectCount');
        projectCount += 1;
        const project = new Project(name, id);
        localStorageData.setDataIntoLocalStorage(name, project);
        localStorageData.setDataIntoLocalStorage('projectCount', projectCount);
        projectsArray.push(project);
        localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
        return project;
      },
      update(id, updates) {
        const oldProjectName = projectsArray.find(x => x.id === id).name;
        const oldProject = localStorageData.getDataFromLocalStorage(oldProjectName);
        const project = localStorageData.getDataFromLocalStorage(oldProject.name);
        const index = projectsArray.findIndex(x => x.name === project.name);
        project.name = updates.name;
        project.todos = oldProject.todos.slice();
        projectsArray.splice(index, 1, project);
        localStorageData.setDataIntoLocalStorage(project.name, project);
        localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
        localStorageData.removeDataFromLocalStorage(oldProject.name);
        return project;
      },
      delete(id) {
        const projectsArray = localStorageData.getDataFromLocalStorage('projectsArray');
        const projectCount = localStorageData.getDataFromLocalStorage('projectCount');
        const todosArray = localStorageData.getDataFromLocalStorage('todosArray');
        const projectInArray = projectsArray.find(project => project.id === id);
        const projectStandAlone = localStorageData.getDataFromLocalStorage(projectInArray.name);
        // Delete all project todos
        const newTodosArray = todosArray.filter(todo => todo.projectID !== projectStandAlone.id);
        localStorageData.setDataIntoLocalStorage('todosArray', newTodosArray);
        // Delete stand alone project
        localStorageData.removeDataFromLocalStorage(projectInArray.name);
        // Delete project from projectsArray
        const newProjectsArray = projectsArray.filter(project => project.id !== projectInArray.id);
        localStorageData.setDataIntoLocalStorage('projectsArray', newProjectsArray);
        // Reduce project count in localstorage by 1
        localStorageData.setDataIntoLocalStorage('projectCount', projectCount - 1);
      },
    };
  }
)();


export default projectsController;
