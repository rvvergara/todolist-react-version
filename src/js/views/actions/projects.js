import uuid from 'uuid';

const getProjects = projects => ({
  type: 'GET_PROJECTS',
  projects,
});

const addProject = ({ name = 'PROJECT', description = 'SOME PROJECT' } = {}) => ({
  type: 'ADD_PROJECT',
  project: {
    id: uuid(),
    name,
    description,
    todos: [],
  },
});

const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id,
  updates,
});

const deleteProject = id => ({
  type: 'DELETE_PROJECT',
  id,
});

export {
  getProjects,
  addProject,
  updateProject,
  deleteProject,
};
