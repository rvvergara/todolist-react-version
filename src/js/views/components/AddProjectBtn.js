import React from 'react';
import PropTypes from 'prop-types';

const AddProjectBtn = (props) => {
  const {
    addProjectMode,
    clickAddProjectBtn,
  } = props;
  const displayClass = !addProjectMode ? 'btn btn-primary mt-1' : 'd-none';
  return (
    <button
      type="button"
      className={displayClass}
      onClick={clickAddProjectBtn}
    >
    Add New Project
    </button>
  );
};

AddProjectBtn.propTypes = {
  addProjectMode: PropTypes.bool.isRequired,
  clickAddProjectBtn: PropTypes.func.isRequired,
};

export default AddProjectBtn;
