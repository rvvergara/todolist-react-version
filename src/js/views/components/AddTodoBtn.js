import React from 'react';
import PropTypes from 'prop-types';

const AddTodoBtn = (props) => {
  const {
    handleTodoBtn,
  } = props;

  return (
    <button
      type="button"
      className="btn btn-primary btn-block"
      onClick={handleTodoBtn}
    >
    Add Todo
    </button>
  );
};

AddTodoBtn.propTypes = {
  handleTodoBtn: PropTypes.func.isRequired,
};

export default AddTodoBtn;
