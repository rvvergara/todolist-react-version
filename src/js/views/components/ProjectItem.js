import React from 'react';
import ProjectsForm from './ProjectsForm';
class ProjectItem extends React.Component {
  state = {
    todos: [],
    editProjectMode: false,
  }

  clickUpdateProjectBtn = () => {
    this.setState(() => ({
      editProjectMode: true,
    }));
  }

  submitProjectForm = (e) => {
    e.preventDefault();
    const id = Number(e.target.getAttribute('data-id'));
    this.props.updateProjectForm(id);
    this.setState(() => ({
      editProjectMode: false,
    }));
    e.target.reset();
  }

  render(){
    const {
      name,
      dataID,
      deleteProject,
      selectProject,
    } = this.props;
    const regular = (
      <div onClick={selectProject}>
        <span
          className="mx-1"
        >
          { name }
        </span>
        <button
          className="btn btn-sm btn-danger"
          id={`delete-proj-${dataID+1}`}
          onClick={deleteProject}
        >
          Delete
        </button>
        <button
          className="btn btn-sm btn-warning"
          onClick={this.clickUpdateProjectBtn}
        >
        Update
        </button>
      </div>
    );
    const editMode = (
      <ProjectsForm
      dataID={dataID+1}
      editProjectMode={this.state.editProjectMode}
      name={name}
      submitProjectForm={this.submitProjectForm}
      />
    )
    if (this.state.editProjectMode){
      return editMode
    }else{
      return regular
    }
  }
};

export default ProjectItem;