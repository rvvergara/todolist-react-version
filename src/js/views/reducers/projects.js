export default (state = [], action) => {
  switch (action.type) {
    case 'GET_PROJECTS':
      return [...state, ...action.projects]

    case 'ADD_PROJECT':
    {
      const index = state.findIndex(project => project.name === action.project.name);
      return index === -1 ? state.concat(action.project) : state;
    }

    case 'UPDATE_PROJECT':
    {
      const { name } = action.updates;
      const conflictIndex = state.findIndex(project => project.name === name);
      const projectIndex = state.findIndex(project => project.id === action.id);
      const updatedProject = { ...state[projectIndex], ...action.updates };
      return conflictIndex === -1 ? [...state.slice(0, projectIndex), updatedProject, ...state.slice(projectIndex + 1)] : state;
    }
    case 'DELETE_PROJECT':
      return state.filter(project => project.id !== action.id);
    default:
      return state;
  }
};
