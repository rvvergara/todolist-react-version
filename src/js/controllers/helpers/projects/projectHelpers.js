import uuid from 'uuid';

import projectsController from '../../projectsController';

import * as localStorageData from '../common/storage';

export const createProjectsArray = () => {
  if (localStorageData.getDataFromLocalStorage('projectsArray') === null) {
    const projectsArray = [];
    localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
  }
};

// Create a default project for new users or after localstorage is cleared
export const createDefaultProject = () => {
  if (localStorageData.getDataFromLocalStorage('projectCount') === 0) {
    const defaultProject = projectsController.create('Default Project', uuid());
  }
  if (localStorageData.getDataFromLocalStorage('Default Project') !== null) {
    const defaultProject = localStorageData.getDataFromLocalStorage('Default Project');
  }
};

// Adding event listener to Project Form
// document.getElementById('projectNameForm').addEventListener('submit', e => {
//   e.preventDefault();
//   submitProjectForm(e.target);
// });
// // Adding event listener to Add Project Btn
// document.getElementById('addProjBtn').addEventListener('click', e => {
//   const action = 'new';
//   showProjectForm(e.target, action);
// });
