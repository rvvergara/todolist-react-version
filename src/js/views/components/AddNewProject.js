import React from 'react';
import { connect } from 'react-redux';
import AddProjectBtn from './AddProjectBtn';
import ProjectsForm from './ProjectsForm';
import {
  addProject,
} from '../actions/projects';
import {
  selectProject
} from '../actions/selectedProject';
import { addProjectModeSwitch, editProjectModeSwitch } from '../actions/projectForm';
import { addTodoModeSwitch, editTodoModeSwitch } from '../actions/todoForm';
import projectsController from '../../controllers/projectsController';

export class AddNewProject extends React.Component {
  state = {
    addProjectMode: false,
    error: '',
    projectName: '',
  };

  handleClickAddBtn = () => {
    this.props.addProjectModeSwitch();
    if(this.props.editProjectMode) this.props.editProjectModeSwitch();
    if(this.props.addTodoMode) this.props.addTodoModeSwitch();
    if(this.props.editTodoMode) this.props.editTodoModeSwitch();
  };

  handleChange = (key, val) => this.setState({
    [key]: val,
  });

  handleSubmitProjectForm = (e) => {
    e.preventDefault();
    const { addProject } = this.props;
    const name = this.state.projectName;
    const isValid = this.validateName(name);
    if(isValid && name){
      this.setState({
        projectName: '',
        error: '',
      });
      this.props.addProjectModeSwitch();
      const newProject = addProject({name});
      projectsController.create(name, newProject.project.id);
      this.props.selectProject(newProject.project);
      this.props.addNewProject(newProject.project);
    } else {
      this.setState({
        error: 'You have entered a duplicate or invalid name'
      });
    };
  };

  validateName = name => {
    const {projects} = this.props;
    return projects.findIndex(project => project.name === name) === -1; 
  };

  render() {
    return (
      <div>
        {!this.props.addProjectMode && <AddProjectBtn
          addProjectMode={this.state.addProjectMode}
          clickAddProjectBtn={this.handleClickAddBtn}
        />}
        { this.props.addProjectMode && <ProjectsForm
          addProjectMode={this.state.addProjectMode}
          submitProjectForm={this.handleSubmitProjectForm}
          handleChange={this.handleChange}
          error={this.state.error}
          name={this.state.projectName}
        />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  projects: state.projects,
  addProjectMode: state.projectForm.addProjectMode,
  editProjectMode: state.projectForm.editProjectMode,
  addTodoMode: state.todoForm.addTodoMode,
  editTodoMode: state.todoForm.editTodoMode,
});

export default connect(mapStateToProps, { addProject, addProjectModeSwitch, editProjectModeSwitch, selectProject, addTodoModeSwitch, editTodoModeSwitch})(AddNewProject);