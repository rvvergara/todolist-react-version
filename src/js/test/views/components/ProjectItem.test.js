import React from 'react';
import { shallow } from 'enzyme';
import { ProjectItem } from '../../../views/components/ProjectItem';
import projectsController from '../../../controllers/projectsController';
import projects from '../../fixtures/projects';

describe('ProjectItem', () => {
  let wrapper;
  let addProjectModeSwitch;
  let addTodoModeSwitch;
  let deleteProject;
  // let deleteTodo;
  let deleteAllTodos;
  let editTodoModeSwitch;
  let editProjectModeSwitch;
  let handleDeleteProject;
  let selectProject;
  let setProjectForEdit;
  let updateProject;

  beforeEach(() => {
    addProjectModeSwitch = jest.fn();
    addTodoModeSwitch = jest.fn();
    deleteProject = jest.fn();
    deleteAllTodos = jest.fn();
    editProjectModeSwitch = jest.fn();
    editTodoModeSwitch = jest.fn();
    handleDeleteProject = jest.fn();
    selectProject = jest.fn();
    setProjectForEdit = jest.fn();
    updateProject = jest.fn();
    wrapper = shallow(
      <ProjectItem
        project={projects[0]}
        addProjectMode={false}
        addTodoMode={false}
        addTodoModeSwitch={addTodoModeSwitch}
        deleteAllTodos={deleteAllTodos}
        editProjectMode={false}
        editTodoMode={false}
        editTodoModeSwitch={editTodoModeSwitch}
        addProjectModeSwitch={addProjectModeSwitch}
        editProjectModeSwitch={editProjectModeSwitch}
        selectProject={selectProject}
        deleteProject={deleteProject}
        setProjectForEdit={setProjectForEdit}
        projectBeingEdited={undefined}
        updateProject={updateProject}
        handleDeleteProject={handleDeleteProject}
        projectTodos={[]}
      />,
    );
  });
  test('it should render correctly without projectsForm', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call selectProject upon click', () => {
    wrapper.find('div').at(0).simulate('click');
    expect(selectProject).toHaveBeenLastCalledWith(projects[0]);
  });

  describe('interacting with ProjectItemBtns clickUpdateProjectBtn', () => {
    test('should call editProjectModeSwitch', () => {
      wrapper.find('ProjectItemBtns').prop('clickUpdateProjectBtn')({
        stopPropagation: () => {},
      });
      expect(editProjectModeSwitch).toHaveBeenCalled();
    });

    test('should call setProjectForEdit', () => {
      wrapper.find('ProjectItemBtns').prop('clickUpdateProjectBtn')({
        stopPropagation: () => {},
      });
      expect(setProjectForEdit).toHaveBeenLastCalledWith(projects[0].id);
    });

    test('should call addProjectModeSwitch if addProjectMode is true', () => {
      wrapper.setProps({ addProjectMode: true });
      wrapper.find('ProjectItemBtns').prop('clickUpdateProjectBtn')({
        stopPropagation: () => {},
      });
      expect(addProjectModeSwitch).toHaveBeenCalled();
    });

    test('should call addTodoModeSwitch if addTodoMode is true', () => {
      wrapper.setProps({ addTodoMode: true });
      wrapper.find('ProjectItemBtns').prop('clickUpdateProjectBtn')({
        stopPropagation: () => {},
      });
      expect(addTodoModeSwitch).toHaveBeenCalled();
    });

    test('should call editTodoModeSwitch if editTodoMode is true', () => {
      wrapper.setProps({ editTodoMode: true });
      wrapper.find('ProjectItemBtns').prop('clickUpdateProjectBtn')({
        stopPropagation: () => {},
      });
      expect(editTodoModeSwitch).toHaveBeenCalled();
    });
  });


  describe('interaction with ProjectItemBtns handleDeleteProject', () => {
    beforeEach(() => {
      projectsController.delete = jest.fn();
    });

    test('should call on deleProject and projectsController.delete', () => {
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(deleteProject).toHaveBeenLastCalledWith(projects[0].id);
    });

    test('should call on projectController.delete action', () => {
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(projectsController.delete).toHaveBeenLastCalledWith(projects[0].id);
    });

    test('should call on deleteAllTodo', () => {
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(deleteAllTodos).toHaveBeenCalled();
    });

    test('should call addProjectModeSwitch if addProjectMode is true', () => {
      wrapper.setProps({ addProjectMode: true });
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(addProjectModeSwitch).toHaveBeenCalled();
    });

    test('should call editProjectModeSwitch if editProjectMode is true', () => {
      wrapper.setProps({ editProjectMode: true });
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(editProjectModeSwitch).toHaveBeenCalled();
    });

    test('should call adddTodoModeSwitch if addTodoMode is true', () => {
      wrapper.setProps({ addTodoMode: true });
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(addTodoModeSwitch).toHaveBeenCalled();
    });

    test('should call editTodoModeSwitch if editTodoMode is true', () => {
      wrapper.setProps({ editTodoMode: true });
      wrapper.find('ProjectItemBtns').prop('handleDeleteProject')();
      expect(editTodoModeSwitch).toHaveBeenCalled();
    });
  });

  describe('interaction with ProjectsForm', () => {
    beforeEach(() => {
      wrapper.setProps({ editProjectMode: true, addProjectMode: false, projectBeingEdited: projects[0].id });
    });
    test('it should render with ProjectsForm', () => {
      expect(wrapper).toMatchSnapshot();
    });
    test('it should handle change in project name input', () => {
      wrapper.find('ProjectsForm').prop('handleChange')('projectName', 'My Project');
      expect(wrapper.state('projectName')).toBe('My Project');
    });
    test('should handle project updating', () => {
      const name = 'Cool Project';
      projectsController.update = jest.fn();
      wrapper.find('ProjectsForm').prop('handleChange')('projectName', name);
      wrapper.find('ProjectsForm').prop('submitProjectForm')({
        preventDefault: () => {},
        target: { reset: () => {} },
      });
      expect(updateProject).toHaveBeenLastCalledWith(projects[0].id, { name });
      expect(projectsController.update).toHaveBeenLastCalledWith(projects[0].id, { name });
      expect(editProjectModeSwitch).toHaveBeenCalled();
      expect(selectProject).toHaveBeenLastCalledWith({ ...projects[0], name });
    });
  });
});
