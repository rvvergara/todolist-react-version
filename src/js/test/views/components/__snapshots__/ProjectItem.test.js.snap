// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectItem interaction with ProjectsForm it should render with ProjectsForm 1`] = `
<ProjectsForm
  handleChange={[Function]}
  name="First"
  submitProjectForm={[Function]}
/>
`;

exports[`ProjectItem it should render correctly without projectsForm 1`] = `
<div
  onClick={[Function]}
>
  <span
    className="mx-1"
  >
    First
  </span>
  <ProjectItemBtns
    clickUpdateProjectBtn={[Function]}
    handleDeleteProject={[Function]}
    project={
      Object {
        "description": "Enim dolore quis tempor sit.",
        "id": 0,
        "name": "First",
        "todos": Array [
          Object {
            "id": 10,
          },
          Object {
            "id": 20,
          },
          Object {
            "id": 30,
          },
        ],
      }
    }
  />
</div>
`;
